### ISSUE TITLE
The specification should maybe say where code lenses are shown in the client

### ISSUE THREAD

**Issue by michaelpj** [microsoft/language-server-protocol#1558]
At the moment, the specification says nothing about how code lenses are displayed. In all the existing implementations that I know of, code lenses are displayed inline in the code using the title of the command to provide the text. But there is not agreement about _where_ they are displayed! For example, VSCode shows them _above_ the target line, and recent `lsp-mode` shows them at the _end_ of the target line.

This is quite important: since code lenses appear _in_ the code, how they are interpreted by the user depends on where they appear!
For example, the Haskell language server provides code lenses that show the type signature for functions that do not have one, inserting them on click. This works perfectly when the code lenses appear _above_ the line, but look very weird if they appear at the _end_.

So it would be good if the specification could provide some guidance for server implementors. Even if the guidance is "you should not make any assumptions about where or whether code lenses are displayed" that would be very useful: then we would know that what we are doing in HLS is making wrong assumptions and we should stop! Basically I would like to know whether it's HLS or the `lsp-mode` maintainers who are doing something wrong :)

**Comment by ericdallo** [comment]
Example of lens on the end of the line on clojure

![2022-10-03_14-19](https://user-images.githubusercontent.com/7820865/193639604-af984c59-e43f-452e-b954-4f0bbb1c0a35.png)

**Comment by rchl** [comment]
Don't know if it brings anything to the table but FYI, Sublime Text has two ways of showing Code Lenses. One that places code lens under the relevant code and one that places it at the far-right of the line:

https://lsp.sublimetext.io/features/#code-lenses

IMO both work for the case pictured there at least.

**Comment by ericdallo** [comment]
BTW It's pretty common to see line flicker on vscode/emacs when code lens are above, but not on emacs when code lens are far-right

**Comment by olegtk** [comment]
I think it's pretty fundamental LSP principle that servers don't control client UI so the guidance should indeed be "you should not make any assumptions about where or whether code lenses are displayed". It's basically implied, just like with any other feature like completion or code actions.

**Comment by dbaeumer** [comment]
I agree with @olegtk that the specification should not suggest any you implementation. What we could do is the following:

- add a client capability that informs the server of where the client shows code lenses (e.g. above, below, far right, right, ...)
- if the capability is missing then "you should not make any assumptions about where or whether code lenses are displayed" is implied.

**Comment by michaelpj** [comment]
Or maybe if servers want to make assumptions about location they should use inlay hints now. I don't know whether inlay hints can replicated the "display on line above" behaviour of code lenses in vscode, though.
